crates/client/data-availability/src/lib.rs

```diff
@@ -77,61 +74,53 @@ impl<B> DataAvailabilityWorkerProving<B>
-                log::error!("db err: {db_err}");
-            };
-
-            match da_mode {
-                DaMode::Validity => {
-                    // Submit the Starknet OS PIE
-                    // TODO: Validity Impl
-                    // run the Starknet OS with the Cairo VM
-                    // extract the PIE from the Cairo VM run
-                    // pass the PIE to `submit_pie` and zip/base64 internal
-                    if let Ok(job_resp) = sharp::submit_pie("TODO") {
-                        log::info!("Job Submitted: {}", job_resp.cairo_job_key);
-                        // Store the cairo job key
-                        if let Err(db_err) = madara_backend.da().update_cairo_job(&block_hash, job_resp.cairo_job_key) {
-                            log::error!("db err: {db_err}");
-                        };
-                    }
-                }
-                _ => {
-                    log::info!("don't prove in remaining DA modes")
+        match da_mode {
+            DaMode::Validity => {
+                // Submit the Starknet OS PIE
+                // TODO: Validity Impl
+                // run the Starknet OS with the Cairo VM
+                // extract the PIE from the Cairo VM run
+                // pass the PIE to `submit_pie` and zip/base64 internal
+                if let Ok(job_resp) = sharp::submit_pie("TODO") {
+                    log::info!("Job Submitted: {}", job_resp.cairo_job_key);
+                    // Store the cairo job key
+                    if let Err(db_err) = madara_backend.da().update_cairo_job(&block_hash, job_resp.cairo_job_key) {
+                        log::error!("db err: {db_err}");
+                    };
                 }
             }
+            _ => {
+                log::info!("don't prove in remaining DA modes")
+            }
         }
     }
 }
 
-impl<B, C, H> DataAvailabilityWorker<B, C, H>
+impl<B, H> DataAvailabilityWorker<B, H>
 where
     B: BlockT,
-    C: ProvideRuntimeApi<B>,
-    C::Api: StarknetRuntimeApi<B>,
-    C: BlockchainEvents<B> + 'static,
-    C: HeaderBackend<B>,
     H: HasherT,
 {
     pub async fn update_state(
         da_client: Box<dyn DaClient + Send + Sync>,
-        client: Arc<C>,
+        mut state_diffs_rx: mpsc::Receiver<BlockDAData>,
         madara_backend: Arc<mc_db::Backend<B>>,
     ) {
-        let mut notification_st = client.import_notification_stream();
+        while let Some(BlockDAData(starknet_block_hash, csd, num_addr_accessed)) = state_diffs_rx.next().await {
+            log::info!(
+                "received state diff for block {starknet_block_hash}: {csd:?}. {num_addr_accessed} addresses accessed."
+            );
 
-        while let Some(notification) = notification_st.next().await {
+            // store the state diff
+            if let Err(db_err) = madara_backend
+                .da()
+                .store_state_diff(&starknet_block_hash, state_diff_to_calldata(csd, num_addr_accessed))
```

Comment:
The value returned by `state_diff_to_calldata(csd, num_addr_accessed)` is read again from db on line 150: 
`DaMode::Sovereign => match madara_backend.da().state_diff(&starknet_block_hash)`

It feels like a waste. It can be kept in memory, in a variable an used directly from there.

Reply:
Does it even need to be stored?
It is the only place this is used. Nobody else is reading it

Reply:
It will be read again during proving. We can however pass it directly as well as now proving is called from `update_state`. However, I am not sure how safe just in memory would be during node restarts or crashes.

Reply:
Will the `prove` task need the encoded stated diff (the payload expected by l1 contract) or the `ThinkStateDiff`?
We are currently storing the former, but I think the latter is what we need.

Reply:
I am not sure, maybe @drspacemn can comment on this. Either ways, I think it will be easy to switch to the latter if needed later on if we are sure we need to store it right now.